README.md
# ajustes-a-direito

CHANGELOG.md
# Change Log
All notable [changes](http://keepachangelog.com/) to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0][unreleased]
### Changed
- Vagrant file box: from ``ubuntu/trusty64``to ``phusion/ubuntu-14.04-amd64`` (run ``vagrant destroy`` and ``vagrant up`` to refresh your setup).
- Code moved to ``/src``.

### Added
- Config file: ``/config.js``.
- Documents directory: ``/docs``.
- NPM package file: ``package.json`` (run ``npm install`` inside ``/vagrant`` to download dependencies).

### Fixed
- Auto start docker containers.

### Removed
- ``/node_modules`` from versioning.


test/test-bitly.js
var config = require('../config/config.js');
var bitlyapi = require('node-bitlyapi');

var Bitly = new bitlyapi({
  client_id: config.bitly.clientId,
  client_secret: config.bitly.clientSecret
});

Bitly.setAccessToken(config.bitly.AccessToken);

Bitly.shortenLink('http://github.com/nkirby/node-bitlyapi', function(err, results) {

  results = JSON.parse(results);

  if (results.status_code !== 200)
    console.log('Error: ' + results.status_txt);
  else
    console.log(results.data.url);

});


test/test-facebook.js
var config = require('../config/config.js');
var fb = require('fb');

fb.setAccessToken(config.facebook.accessToken);

var body = 'A\n\nB\n&nbsp;\nC';

fb.api(config.facebook.pageId + '/feed', 'post', { message: body }, function (res) {
  if(!res || res.error) {
    console.log(!res ? 'error occurred' : res.error);
    return;
  }
  var tokens = res.id.split(/_/g);
  console.log('https://www.facebook.com/AjustesDireitos/posts/' + tokens[1]);
});


test/populate-queue.js
var config = require('../config/config.js');

var fivebeans = require('fivebeans');

var entries = [{
  "ContracObject": "Aluguer de oito (8) autocarros para assegurar transportes escolares e o serviço de vigilância humana para acompanhamento e supervisão dos alunos nos trajetos do transporte escolar",
  "Buyer": "Camara Municipal de Palmela (506187543)",
  "Seller": "\"Traviama - Transportes Terrestres de Passageiros, Lda (506855155)\"",
  "Cost": "\"173.370,40 €\"",
  "Date": "27-10-2015",
  "Hash": "f201e03ca0b888b25e90db36103f6ae1"
}, {
  "ContracObject": "Construção das Piscinas Cobertas de Pinhel",
  "Buyer": "Município de Pinhel (506787249)",
  "Seller": "\"Biosfera Construções, Unip. Lda. (508307880)\"",
  "Cost": "\"1.562.000,00 €\"",
  "Date": "27-10-2015",
  "Hash": "be79edb5e91d2cc63481ad9a00b18449"
}, {
  "ContracObject": "\"Prestação de serviços de manutenção de espaços exteriores, incluindo as zonas verdes e pavimentadas com calçada dos passeios e pracetas\"",
  "Buyer": "Município de Palmela (506187543)",
  "Seller": "\"Centro de Jardinagem Sogrene, Lda (502778687)\"",
  "Cost": "\"152.042,71 €\"",
  "Date": "27-10-2015",
  "Hash": "b60c261d6bb9184500a2c505dea7845b"
}, {
  "ContracObject": "Fornecimento de equipamentos audiovisuais e de ativos de rede e equipamentos de comunicações para o Centro de Documentação 25 de Abril e Centro de Estudos Sociais da Universidade de Coimbra",
  "Buyer": "Universidade de Coimbra (501617582)",
  "Seller": "\"Decunify - Soluções e Comunicações, SA (504889893)\"",
  "Cost": "\"99.493,35 €\"",
  "Date": "27-10-2015",
  "Hash": "933ac5decaf717bf38f940c49c8f0693"
}, {
  "ContracObject": "\"União das Freguesias de Santiago de Litém, S. Simão de Litém e Albergaria dos Doze / Asfaltagem de estradas e caminhos na Freguesia [E.M. 532-1 (Castelo - Albergaria dos Doze), Albergaria dos Doze, Cartaria e Vidoeira] - Proc. n.º 49/2015\"",
  "Buyer": "Município de Pombal (506334562)",
  "Seller": "\"Contec - Construção e Engenharia, S.A. (501436162)\"",
  "Cost": "\"122.554,28 €\"",
  "Date": "27-10-2015",
  "Hash": "3cb4ddff1975c6d7acdd58907ab6475e"
}, {
  "ContracObject": "\"Fornecimento de energia elétrica para as instalações alimentadas em MT – Média Tensão (lote 1), BTE - Baixa Tensão Especial (lote 2), BTN - Baixa Tensão Normal “Iluminação Pública” (lote 3), e BTN- Baixa Tensão Normal, excluindo Iluminação Pública – Instalações Municipais (lote 4), pelo período de um ano.\"",
  "Buyer": "Município de Braga (506901173)",
  "Seller": "\"EDP Comercial – Comercialização de Energia, S.A. (503504564)\"",
  "Cost": "\"2.238.649,37 €\"",
  "Date": "27-10-2015",
  "Hash": "e6d08cff1770a052133e5d5105d1ad2c"
}, {
  "ContracObject": "Prestação de Serviços de Limpeza Química das Caldeiras",
  "Buyer": "\"Valorsul - Valorização e Tratamento de Resíduos Sólidos das Regiões de Lisboa e do Oeste, S. A. (509479600)\"",
  "Seller": "\"LQI – Serviços Industrias, S.A. (501438114)\"",
  "Cost": "\"0,00 €\"",
  "Date": "27-10-2015",
  "Hash": "ab257355dc411b9143c2522791ac0ad2"
}, {
  "ContracObject": "Aquisição de Mobiliário Geral e Mobiliário Administrativo e Hospitalar para o Lar de Idosos e Creche",
  "Buyer": "Centro Social Professora Elisa Barros Silva (502023910)",
  "Seller": "\"Móvel4 II Companhia do Móvel, Lda. (510655238)\"",
  "Cost": "\"144.106,00 €\"",
  "Date": "27-10-2015",
  "Hash": "ddfcc86bb3e83dc010f7b81b04cec2a5"
}, {
  "ContracObject": "Prestação de serviços de manutenção do equipamento lúdico-desportivo nas Escolas do 1.º Ciclo do Ensino Básico do Concelho de Almada",
  "Buyer": "Câmara Municipal de Almada (500051054)",
  "Seller": "\"PLAY PLANET - MOB. URBANO, CONSTRUÇÃO E PAISAGISMO, LDA (509295770)\"",
  "Cost": "\"25.000,00 €\"",
  "Date": "27-10-2015",
  "Hash": "6b41da0c5289551001e006717d99cd7b"
}, {
  "ContracObject": "\"Limpeza e Beneficiação dos Rios e Ribeiros no Concelho (Reabilitação, Limpeza e valorização do Rio Arunca) - Proc. n.º 39/2015\"",
  "Buyer": "Município de Pombal (506334562)",
  "Seller": "\"MarvãoMáquinas - Aluguer de Máquinas, Lda. (503623130)\"",
  "Cost": "\"120.682,46 €\"",
  "Date": "27-10-2015",
  "Hash": "64004c3252c50b5f0bf8b7b31597b698"
}, {
  "ContracObject": "\"Fornecimento de energia eléctrica em Média Tensão, Baixa Tensão Especial, Baixa Tensão Normal e Iluminação Pública. \"",
  "Buyer": "Município de Pinhel (506787249)",
  "Seller": "\"HEN - Serviços Energéticos, Lda. (510287050)\"",
  "Cost": "\"573.029,67 €\"",
  "Date": "27-10-2015",
  "Hash": "e1b8332b502882c00b609453d6159e59"
}, {
  "ContracObject": "Aquisição de sobresselentes panhard m11",
  "Buyer": "Estado Maior do Exército (600021610)",
  "Seller": "Panhard General Defense (25652036591)",
  "Cost": "\"349.686,55 €\"",
  "Date": "26-10-2015",
  "Hash": "e5084da2eb13f5c2fcf98d4bf49cff24"
}, {
  "ContracObject": "Fornecimento contínuo de gasóleo rodoviário .",
  "Buyer": "Município de Campo Maior (501175229)",
  "Seller": "\"Francisco Laia Nunes, Lda. (502659726)\"",
  "Cost": "\"126.100,00 €\"",
  "Date": "26-10-2015",
  "Hash": "26f729b70c508908a756b92ced34cf3f"
}, {
  "ContracObject": "Serviços de limpeza para instalações da C.M.P.",
  "Buyer": "Município de Palmela (506187543)",
  "Seller": "\"Euromex - Facility Services, Lda (502629428)\"",
  "Cost": "\"163.440,00 €\"",
  "Date": "26-10-2015",
  "Hash": "18bced480d5d4d0592261a20173a74a2"
}, {
  "ContracObject": "\"Fornecimento de chapas para tromel tr01,02,03,04 \"",
  "Buyer": "\"Tratolixo - Tratamento de Resíduos Sólidos, E. I. M. (502444010)\"",
  "Seller": "\"Masias Recycling, S.L. (B-17282567)\"",
  "Cost": "\"90.815,76 €\"",
  "Date": "26-10-2015",
  "Hash": "9801e1554307e90df4faf7b651898f20"
}, {
  "ContracObject": "Serviços de limpeza e desobstrução de emissários e transporte de lamas líquidas",
  "Buyer": "\"Simlis - Saneamento integrado dos Municípios do Lis, S. A. (504864688)\"",
  "Seller": "\"Octalimpa - Limpezas Unipessoal, Lda. (507187482)\"",
  "Cost": "\"231.000,00 €\"",
  "Date": "26-10-2015",
  "Hash": "0331beb377f8cd3a20f06bbb356d32df"
}, {
  "ContracObject": "\"Recolha e Tratamento de lamas digeridas da Central de Digestão Anaeróbia, na Abrunheira, Mafra\"",
  "Buyer": "\"Tratolixo - Tratamento de Resíduos Sólidos, E. I. M. (502444010)\"",
  "Seller": "\"CITRI, Centro Integrado de Tratamento de Resíduos Industriais, S.A. (504472046)\"",
  "Cost": "\"65.520,00 €\"",
  "Date": "27-10-2015",
  "Hash": "73951e213bb8e06f734cfb471e202095"
}, {
  "ContracObject": "Deposição de monstros",
  "Buyer": "\"Tratolixo - Tratamento de Resíduos Sólidos, E. I. M. (502444010)\"",
  "Seller": "\"CITRI, Centro Integrado de Tratamento de Resíduos Industriais, S.A. (504472046)\"",
  "Cost": "\"91.960,00 €\"",
  "Date": "27-10-2015",
  "Hash": "b2f842464bb4403c179937444771d4cf"
}, {
  "ContracObject": "\"PRESTAÇÃO DE SERVIÇOS DE MANUTENÇÃO E ASSISTÊNCIA TÉCNICA AOS SISTEMAS AVAC E INSTALAÇÕES DE AR CONDICIONADO DE CENTROS DE SAÚDE E SERVIÇOS DO SESARAM, EPE.\"",
  "Buyer": "\"Serviço de Saúde da Região Autónoma da Madeira, E. P. E. (511228848)\"",
  "Seller": "\"OPENLINE FACILITY SERVICES, SA. (508622069)\"",
  "Cost": "\"41.600,00 €\"",
  "Date": "27-10-2015",
  "Hash": "9d0f664027190a8e57f9d221b936f6eb"
}, {
  "ContracObject": "\"Aquisição de reagentes para a realização de análises clínicas nas unidades funcionais da química clínica e urgência do Serviço de Patologia Clínica do Centro Hospitalar Lisboa Norte, EPE\"",
  "Buyer": "\"Centro Hospitalar Lisboa Norte, E. P. E. (508481287)\"",
  "Seller": "\"Roche, Lda. (504282921)\"",
  "Cost": "\"1.355.401,33 €\"",
  "Date": "27-10-2015",
  "Hash": "7da3f5caed29dff294150d9e49a97540"
}, {
  "ContracObject": "\"Execução de moviemntação de terras, demolições, pavimentos e colocação de lancis, colocação de mobiliário urbano, arborização em caldeiras e criação de zona verde em prado florido incluindo rede de rega.\"",
  "Buyer": "Município de Palmela (506187543)",
  "Seller": "\"Florindo Rodrigues Júnior & Filhos, SA (500603456)\"",
  "Cost": "\"39.598,98 €\"",
  "Date": "27-10-2015",
  "Hash": "62a81e33b8bb4061c4cbce54e19b62a5"
}, {
  "ContracObject": "\"Contrato de fiscalização e controlo da empreitada de Reabilitação da \"\"Ilha\"\" da Bela Vista (1.ª Fase) (SE.004.2015.010)\"",
  "Buyer": "\"CMPH - DomusSocial - Empresa de Habitação e Manutenção do Município do Porto, EEM (505037700)\"",
  "Seller": "\"Cotefis - Gestão de Projectos, SA (502693622)\"",
  "Cost": "\"33.732,50 €\"",
  "Date": "27-10-2015",
  "Hash": "2d6e30067e1296ebb67a94a80d67cb3e"
}, {
  "ContracObject": "\"Empreitada de \"\"Dragagens de manutenção no Portinho de Vila Praia de Âncora\"\"\"",
  "Buyer": "\"Direção-Geral de Recursos Naturais, Segurança e Serviços Marítimos (600084973)\"",
  "Seller": "\"Alexandre Barbosa Borges, S.A. (500553408)\"",
  "Cost": "\"329.867,10 €\"",
  "Date": "27-10-2015",
  "Hash": "f397bff8a4ce03b3cd6a628e39390ab1"
}, {
  "ContracObject": "Colheita e Análises Fisico-Químicas e Microbiológicas para Controlo de Qualidade da Água no Subsistema de Água Residual",
  "Buyer": "\"Águas de Santo André, S. A. (505600005)\"",
  "Seller": "\"Laboratório de Águas do Litoral Alentejano, Lda. (503310980)\"",
  "Cost": "\"86.804,90 €\"",
  "Date": "27-10-2015",
  "Hash": "d621f93060e04b8b19aac5a2804e8ce0"
}, {
  "ContracObject": "\" AQUISIÇÃO DE SISTEMAS DE PURIFICAÇÃO DE ÁGUA PARA OBTENÇÃO DE ÁGUA PURIFICADA DE QUALIDADE, PARA A UNIDADE DE INVESTIGAÇÃO IBIMED DA SECÇÃO AUTÓNOMA DAS CIÊNCIAS DE SAÚDE UNIVERSIDADE DE AVEIRO\"",
  "Buyer": "Universidade de Aveiro (501461108)",
  "Seller": "Millipore S.A.S.  (FR59434691192)",
  "Cost": "\"52.186,58 €\"",
  "Date": "27-10-2015",
  "Hash": "d523d7ce900e4f3412b7802aff9af6ee"
}, {
  "ContracObject": "\"Execução de Infraestruturas de Águas Pluviais na Praceta de S.Bartolomeu, Rua da Bela Vista, Rua do Viso, Rua D.Sancho I e Rua Cidade de Viseu\"",
  "Buyer": "Município de Aveiro (505931192)",
  "Seller": "\"MANUEL FRANCISCO DE ALMEIDA, S.A. (500178585)\"",
  "Cost": "\"57.128,08 €\"",
  "Date": "26-10-2015",
  "Hash": "5da84d14495df64f7b614eeb2047872e"
}, {
  "ContracObject": "\"Fornecimento de Energia Eléctrica para as Instalações Alimentadas em Média Tensão e Baixa Tensão Especial, das Instalações dos Serviços Municipalizados\"",
  "Buyer": "Serviços Municipalizados de Viseu (680020063)",
  "Seller": "\"EDP Comercial - Comercialização de Energia, S.A. (503504564)\"",
  "Cost": "\"593.688,39 €\"",
  "Date": "26-10-2015",
  "Hash": "c18ce8377a6b55d6cab1812131cbdab3"
}, {
  "ContracObject": "CONTRATO DE AQUISIÇÃO DE SERVIÇOS DE RECOLHA DE RESÍDUOS SÓLIDOS URBANOS NO CONCELHO DE VALPAÇOS",
  "Buyer": "Município de Valpaços (506874320)",
  "Seller": "\"SUMA - SERVIÇOS URBANOS E MEIO AMBIENTE, S.A. (503210560)\"",
  "Cost": "\"174.000,00 €\"",
  "Date": "26-10-2015",
  "Hash": "7b0e0f1a2516adb62bc5791aa927cd79"
}, {
  "ContracObject": "Combustíveis a granel (gasolina 95 e gasóleo rodoviário)",
  "Buyer": "\"Inframoura, E.M. (504915266)\"",
  "Seller": "\"NORBAT - Baterias, Combustíveis e Lurbificantes, Lda. (505377934)\"",
  "Cost": "\"143.880,60 €\"",
  "Date": "26-10-2015",
  "Hash": "34d7509d3baa3b22efd77dab55c11548"
}, {
  "ContracObject": "Serviços de inspecção revisão e manutenção e recarga de meios de combate a incêndio",
  "Buyer": "Faculdade de Arquitetura da Universidade de Lisboa (502784083)",
  "Seller": "Visacção - Segurança Privada S.A. (507756002)",
  "Cost": "\"3.612,30 €\"",
  "Date": "26-10-2015",
  "Hash": "08ff71f1941f2bbeef20363d5271f221"
}];

var maxEntriesToProcess = 1;
var processedEntries = 0;
var toProccessEntries = Math.min(entries.length, maxEntriesToProcess);

function populateQueue() {
  for (var i = 0; i < toProccessEntries; i++) {
    processedEntries++;
    beans.put(JSON.stringify(entries[i]));
  }
}

//=====================  FIVE BEANS

var beans = {
  init: function() {
    queueClient = new fivebeans.client(config.beanstalkd.address, config.beanstalkd.port);
    queueClient.on('connect', function() {
        queueClient.use(config.beanstalkd.tube, function(err, response) {
          populateQueue();
        });
      })
      .on('error', function(err) {
        console.log("Error connection Beanstalkd - " + err);
      })
      .on('close', function() {
        console.log("Connection to Beanstalkd closed");
      })
      .connect();
  },

  put: function(payload) {
    return function(payload, inProcessedEntries, inEntriesToProcess) {
      queueClient.put(0, 0, 30, payload, function(err, jobid) {
        if (err)
          console.log("Error inserting in queue - Payload: " + JSON.stringify(payload));

        if (inProcessedEntries === toProccessEntries) {
          queueClient.end();
        }
      });
    }(payload, processedEntries)
  }
}

beans.init();


publisher/src/config.js
var path = require('path');
var winston = require('winston');

var config = {};

config.env = {};
config.env.inProduction = process.env.ENV === 'prd';
config.env.coldDownPeriod = config.env.inProduction // Time in ms to wait between each reserve.
	? 60 * 1000  // Wait 60s in prd.
	: 60 * 1000; // Wait 60s in dev.

config.beanstalkd = {};
config.beanstalkd.port = '11300';
config.beanstalkd.address = 'beanstalkd';
config.beanstalkd.tube = config.env.inProduction ? 'contracts-prod' : 'contracts-dev';

config.twitter = {};
config.twitter.maxChars = 280;
config.twitter.apiKey = '';
config.twitter.apiSecret = '';
config.twitter.bearerToken = '';
config.twitter.accessToken = '';
config.twitter.accessTokenSecret = ''; 
config.twitter.clientId = '';
config.twitter.clientSecret = '';

config.bitly = {};
config.bitly.clientId = '';
config.bitly.clientSecret = '';
config.bitly.AccessToken = '';

config.facebook = {}
config.facebook.pageId = '';
config.facebook.appId = '';
config.facebook.appSecret = '';
config.facebook.accessToken = '';
config.facebook.postUrl = '';

config.logging = {};
config.logging.transports = [];

config.logging.transports.push(
	new (winston.transports.File)({
		level: config.env.inProduction ? 'info' : 'debug',
		filename: 'main-' + path.basename(process.mainModule.filename, path.extname(process.mainModule.filename)) + '.log',
		maxsize: 1048576, // 1 MB
		maxFiles: 10,
		json: false,
		formatter: function(options) {
			return (new Date()).toString() +' ['+ options.level.toUpperCase() +'] '+ (undefined !== options.message ? options.message : '') + (options.meta && Object.keys(options.meta).length ? '\n\t'+ JSON.stringify(options.meta) : '' );
		}
	}));

if (!config.env.inProduction) {
	config.logging.transports.push(
		new (winston.transports.Console)({
			level: 'debug',
			formatter: function(options) {
				return (new Date()).toString() +' ['+ options.level.toUpperCase() +'] '+ (undefined !== options.message ? options.message : '') + (options.meta && Object.keys(options.meta).length ? '\n\t'+ JSON.stringify(options.meta) : '' );
			}
		}));
}

module.exports = config;


publisher/src/app.js
// Publisher

var TwitterApi = require('twitter-api-v2').default;

var util = require('util');
var fb = require('fb');
var bitlyapi = require('node-bitlyapi');
var winston = require('winston');
var fivebeans = require('fivebeans');
var https = require('https');  
var qs = require('querystring');

var config = require('./config.js');

var logger = new(winston.Logger)({
  transports: config.logging.transports
});

/**
 * Entry point.
 */
var main = {
  init: function() {
    logger.log('info', 'Starting publisher, running in PRD: ' + config.env.inProduction);
    
    // Register signal handlers.
    process.on('SIGINT', this.signalHandler);
    process.on('SIGTERM', this.signalHandler);

    // Catches uncaught exceptions.
    process.on('uncaughtException', function(e) {
      logger.log('error', 'Uncaught Exception: ' + e.stack);

      main.finalize(function() {
        logger.log('info', 'Processed exit, code 1');
        process.exit(1);
      });
    });

    // Initialize beanstalkd listener.
    beans.init();
  },

  signalHandler: function() {
    logger.log('info', 'SIGINT received');

    main.finalize(function() {
      logger.log('info', 'Processed exit, code 0');
      process.exit(0);
    });
  },

  finalize: function(callback) {
    beans.endCallback = callback
    beans.client.end();
  }
};

/**
 * Handle beanstalkd connections.
 */
var beans = {

  client: null,

  endCallback: null,

  init: function() {
    beans.client = new fivebeans.client(config.beanstalkd.address, config.beanstalkd.port);

    beans.client.on('connect', function() {

      logger.log('info', 'Beanstalkd connected ' + config.beanstalkd.address + ' ' + config.beanstalkd.port);

      beans.client.use(config.beanstalkd.tube, function(err, response) {
        if (err) {
          logger.log('error', 'Beanstalkd use error: ' + err);
          process.exit(1);
        }

        // Connected, wait for jobs.
        beans.client.watch(config.beanstalkd.tube, function(err, numwatched) {
          if (err) {
            logger.log('error', 'Beanstalkd watch error: ' + err);
            process.exit(1);
          }

          logger.log('info', 'Beanstalkd watching ' + config.beanstalkd.tube);
          beans.reserve();
        });
      });
    });

    beans.client.on('error', function(err) {
      logger.log('error', 'Beanstalkd error: ' + err);
    });

    beans.client.on('close', function() {
      logger.log('info', 'Beanstalkd closed');
      if (beans.endCallback !== undefined) {
        beans.endCallback();
        beans.endCallback = undefined;
      }
    });

    beans.client.connect();
  },

  reserve: function() {
    beans.client.stats_tube(config.beanstalkd.tube, function(err, response) {
      if (err)
        logger.log('error', 'Beanstalkd stats error: ' + err);

      logger.log('info', 'Beanstalkd ready jobs: ' + response['current-jobs-ready']);
    });

    beans.client.reserve(function(err, jobid, payload) {
      if (err) {
        logger.log('error', 'Beanstalkd reserve error: ' + err);
        process.exit(1);
      }

      logger.log('info', 'Beanstalkd reserve: ' + jobid);
      var entry = JSON.parse(payload);

      // Add jobid to payload.
      entry.jobid = jobid;

      // Process payload.
      publish.process(entry);
    });
  },

  destroy: function(jobid) {
    beans.client.destroy(jobid, function(err) {
      logger.log('info', 'Beanstalkd destroy: ' + jobid);
      if (err)
        logger.log('error', 'Beanstalkd destroy error: ' + err);

      // Wait before next reserve.
      setTimeout(function() {
        beans.reserve(); // Reserve next message.
      }, config.env.coldDownPeriod);
    });
  }
};

/**
 * Handle publishing to 3rd party services (e.g.: Twitter, Facebook, etc.).
 */
var publish = {

  process: function(message) {

    logger.log('info', 'Processing: ' + JSON.stringify(message));

    baseAPI.getDetails(message.id, async function(entry) {
        await publish.twitter.post(entry);
        
        // Job processed, delete it.
        beans.destroy(message.jobid);
    });
  },

  facebook: {

    init: function() {
      fb.setAccessToken(config.facebook.accessToken);
    }(),

    post: function(entry, callback) {
      var post = this.createPost(entry);

      logger.log('info', 'Publishing facebook post: ' + (post.length > 40 ? post.substring(0, 40) + '..' : post).replace(/\n/g, ' '));

      if (config.env.inProduction) {
        fb.api(config.facebook.pageId + '/feed', 'post', {
          message: post
        }, function(res) {
          // Post to facebook done.
          entry.FacebookPostUrl = ''; // Default value.

          if (!res || res.error) {
            // Error posting to facebook.
            logger.log('error', 'Publishing facebook post error: ' + (!res ? 'error occurred' : res.message));
          } else {
            // Post to facebook ok.
            var tokens = res.id.split(/_/g);
            if (tokens.length === 2)
              entry.FacebookPostUrl = util.format(config.facebook.postUrl, tokens[1]);
          }

          callback(entry);
        });
      } else {
        // Dev mode.
        entry.FacebookPostUrl = util.format(config.facebook.postUrl, 'DEV-ID');
        callback(entry);
      }
    },

    createPost: function(entry) {
      var post = '\
{CONTRACT}\n\
\n\
Adjudicante: {BUYER}\n\
Fornecedor: {SELLER}\n\
Valor: {COST}\n\
\n\
{TAGS}';

      post = post.replace(/{CONTRACT}/g, utils.cleanText(entry.ContracObject));
      post = post.replace(/{BUYER}/g, utils.cleanText(entry.Buyer));
      post = post.replace(/{SELLER}/g, utils.cleanText(entry.Seller));
      post = post.replace(/{COST}/g, utils.cleanCurrency(entry.Cost));
      post = post.replace(/{TAGS}/g, utils.createCityHashtag(utils.cleanText(entry.ExecLocal), false));

      return post;
    }
  },

  bitly: {
    bitly: null,

    init: function() {
      bitly = new bitlyapi({
        client_id: config.bitly.clientId,
        client_secret: config.bitly.clientSecret
      });

      bitly.setAccessToken(config.bitly.AccessToken);
    }(),

    process: function(entry, callback) {
      logger.log('info', 'Processing bitly url: ' + entry.FacebookPostUrl);

      if (config.env.inProduction) {
        bitly.shortenLink(entry.FacebookPostUrl, function(err, results) {
          entry.ShortFacebookPostUrl = ''; // Default value.

          try {
            results = JSON.parse(results);
            if (results.status_code !== 200) {
              logger.log('error', 'Processing bitly url error: ' + results.status_txt);
            } else {
              entry.ShortFacebookPostUrl = results.data.url;
            }
          } catch (e) {
            logger.log('error', 'Uncaught Exception on bitly.process(): ' + e.stack);
          }

          callback(entry);
        });
      } else {
        // Dev mode.
        entry.ShortFacebookPostUrl = 'http://on.fb.com/XXXXXX';
        callback(entry);
      }
    }
  },

  twitter: {

    post: async function(entry) {

      // Create tweet.
      const tweet = this.createTweet(entry);

      logger.log('info', 'Publishing twitter post with ' + tweet.length + ' chars.');

      if (config.env.inProduction) {

        // Post tweet.
        try {
          const userClient = new TwitterApi({
            appKey: config.twitter.apiKey,
            appSecret: config.twitter.apiSecret,
            accessToken: config.twitter.accessToken,
            accessSecret: config.twitter.accessTokenSecret
          });
  
          const rwClient = userClient.readWrite;
          const { data: createdTweet } = await rwClient.v2.tweet(tweet);
  
          logger.log('info', 'Tweet response: ' + createdTweet.id);
        } catch (e) {
          logger.log('error', 'Uncaught exception on twitter.post(): ' + e.stack);
        }

      } else {

        // Dev mode, do not post to Twitter.
        logger.log('info', 'Tweet response: -- running in dev --');
      }
    },

    createTweet: function(entry) {
      var sep = ['\n'];

      // Clean texts.
      var cost =          '💶 ' + utils.cleanCurrency(entry.initialContractualPrice);
      var contracObject = '🤝 ' + utils.cleanText(entry.objectBriefDescription);
      var seller =        '🛠 Fornecido por: ' + utils.cleanText(entry.contracted[0].description); 
      var buyer =         '🏛️ Para: ' + utils.cleanText(entry.contracting[0].description);
      var hashtag =       '📍 ' + utils.createCityHashtag(utils.cleanText(entry.executionPlace), true);
      var link =          'ℹ️ Detalhes em https://www.base.gov.pt/Base4/pt/detalhe/?type=contratos&id=' + entry.id; //entry.ShortFacebookPostUrl;

      var currentSep, maxVariablePartSize, variablePartSize;
      for (currentSep = 0; currentSep < sep.length; currentSep++) {
        // Check if it is possible to post using the largest separetor first.
        var fixedPart = sep[currentSep] + cost + sep[currentSep] + hashtag + sep[currentSep] + link;
        maxVariablePartSize = config.twitter.maxChars - fixedPart.length;
        variablePartSize = contracObject.length + sep[currentSep].length + seller.length + buyer.length;

        if (variablePartSize <= maxVariablePartSize) {
          return cost + sep[currentSep] + contracObject + sep[currentSep] + seller + sep[currentSep] + buyer + sep[currentSep] + hashtag + sep[currentSep] + link;
        }
      }

      // Not possible to post using the largest separetor, use the smallest.
      currentSep = sep.length - 1;

      if (variablePartSize > maxVariablePartSize) {
        var overhead = variablePartSize - maxVariablePartSize; // Get size in excess.

        while (overhead > 0) {
          if (contracObject.length > buyer.length) {
            var diff = contracObject.length - buyer.length;
            var apply = Math.min(overhead, diff);
            contracObject = utils.trimToSize(contracObject, contracObject.length - apply).trim();
            overhead = overhead - apply;
          } else if (contracObject.length < buyer.length) {
            var diff = buyer.length - contracObject.length;
            var apply = Math.min(overhead, diff);
            buyer = utils.trimToSize(buyer, buyer.length - apply).trim();
            overhead = overhead - apply;
          } else {
            contracObject = utils.trimToSize(contracObject, contracObject.length - (overhead / 2)).trim();
            buyer = utils.trimToSize(buyer, buyer.length - (overhead / 2)).trim();
            overhead = 0;
          }
        }
      }

      return cost + sep[currentSep] + contracObject + sep[currentSep] + seller + sep[currentSep] + buyer + sep[currentSep] + hashtag + sep[currentSep] + link;
    },
  }
};

var baseAPI = {

  getDetails: function (id, callback) {
    var options = {
      'method': 'POST',
      'hostname': 'www.base.gov.pt',
      'path': '/Base4/pt/resultados/',
      'headers': {
        'Host': 'www.base.gov.pt',
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:102.0) Gecko/20100101 Firefox/102.0',
        'Accept': 'text/plain, */*; q=0.01',
        'Accept-Language': 'en-US,en;q=0.5',
        'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        'X-Requested-With': 'XMLHttpRequest',
        'Content-Length': '32',
        'Origin': 'https://www.base.gov.pt',
        'DNT': '1',
        'Connection': 'keep-alive',
        'Referer': 'https://www.base.gov.pt/Base4/pt/detalhe/?type=contratos&id=' + id,
        'Sec-Fetch-Dest': 'empty',
        'Sec-Fetch-Mode': 'cors',
        'Sec-Fetch-Site': 'same-origin',
        'Sec-GPC': '1'
      },
      'maxRedirects': 20
    };
    
    var req = https.request(options, function (res) {
      var chunks = [];
    
      res.on("data", function (chunk) {
        chunks.push(chunk);
      });
    
      res.on("end", function (chunk) {
        callback(JSON.parse(Buffer.concat(chunks)));
      });
    
      res.on("error", function (error) {
        console.error(error);
      });
    });
    
    var postData = qs.stringify({
      'type': 'detail_contratos',
      'id': id
    });
    
    req.write(postData);
    
    req.end();
  }
};

/**
 * Generic stuff.
 */
var utils = {

  trimToSize: function(str, maxLength) {
    if (str.length > maxLength)
      return str.substring(0, maxLength - 2) + '..';

    return str;
  },

  // Remove ", (9999) and multiple spaces.
  cleanText: function(text) {
    text = text.replace(/"/g, '');
    text = text.replace(/\(\d+\)/g, '');
    text = text.replace(/\s+/g, ' ');
    return text.trim();
  },

  // Clean currency entries by removing decimals, separators and multiple spaces.
  cleanCurrency: function(currency) {
    currency = currency.replace(/"/g, '');
    currency = currency.replace(/\'|\s/g, '');
    currency = currency.replace(/,\d\d/g, '');
    currency = currency.replace(/\s+/g, ' ');
    currency = currency.replace(/\./g, '');
    return currency.trim();
  },

  createCityHashtag: function(location, isTweet) {
    try {
      // Ignore empty locations.
      if (location === undefined || location === '')
        return '';

      //array to save cities without duplicates
      var cities = [];

      // Handle multiple locations, only use the first one.
      var locations = location.split(/\n/);
      if (locations.length === 0)
        return '';


      for (var i in locations) {
        location = locations[i];

        // Split location by 'País', 'Distrito', 'Concelho'.
        var tokens = location.split(/,/);
        if (tokens.length !== 3)
          return ''

        for (var j = 1; j < tokens.length; j++) {

          // Convert city to 'CamelCase' and remove spaces.
          var city = tokens[j];
          city = city.replace(/(?:^|\s)\S/g, function(a) {
            return a.toUpperCase();
          });
          city = city.replace(/\s/g, '');

          //returns only one location for twitter - the first location
          if (isTweet)
            return '#' + city;

          if (cities[city] === undefined) {
            cities[city] = '#' + city
          }
        }
      }

      var returnCities = '';
      for (var key in cities)
        returnCities += cities[key] + ' ';

      return returnCities.trim();

    } catch (e) {
      logger.log('error', 'Uncaught Exception on utils.createCityHashtag(): ' + e.stack);
      return '';
    }
  }
};


main.init();
